// @maxhub/max-bot-api/dist/core/helpers/upload.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Upload = void 0;
const fs = __importStar(require("fs"));
const node_crypto_1 = require("node:crypto");
const node_path_1 = __importDefault(require("node:path"));
const api_1 = require("../network/api");
const DEFAULT_UPLOAD_TIMEOUT = 20000; // ms
class Upload {
    constructor(api) {
        this.api = api;
        this.getStreamFromSource = async (source) => {
            if (typeof source === 'string') {
                const stat = await fs.promises.stat(source);
                const fileName = node_path_1.default.basename(source);
                if (!stat.isFile()) {
                    throw new Error(`Failed to upload ${fileName}. Not a file`);
                }
                const stream = fs.createReadStream(source);
                return {
                    stream,
                    fileName,
                    contentLength: stat.size,
                };
            }
            if (source instanceof Buffer) {
                return {
                    buffer: source,
                    fileName: (0, node_crypto_1.randomUUID)(),
                };
            }
            const stat = await fs.promises.stat(source.path);
            let fileName;
            if (typeof source.path === 'string') {
                fileName = node_path_1.default.basename(source.path);
            }
            else {
                fileName = (0, node_crypto_1.randomUUID)();
            }
            return {
                stream: source,
                contentLength: stat.size,
                fileName,
            };
        };
        this.upload = async (type, file, options) => {
            const { url: uploadUrl } = await this.api.raw.uploads.getUploadUrl({ type });
            const uploadController = new AbortController();
            const uploadInterval = setTimeout(() => {
                uploadController.abort();
            }, options?.timeout || DEFAULT_UPLOAD_TIMEOUT);
            try {
                if ('stream' in file) {
                    return await this.uploadFromStream({
                        file,
                        uploadUrl,
                        abortController: uploadController,
                    });
                }
                return await this.uploadFromBuffer({
                    file,
                    uploadUrl,
                    abortController: uploadController,
                });
            }
            finally {
                clearTimeout(uploadInterval);
            }
        };
        this.uploadFromStream = async ({ file, uploadUrl, abortController }) => {
            return new Promise((resolve, reject) => {
                let prevChunk = -1;
                let uploadData = null;
                file.stream.on('data', async (chunk) => {
                    file.stream.pause();
                    const startBite = prevChunk + 1;
                    const endBite = prevChunk + chunk.length;
                    const uploadRes = await fetch(uploadUrl, {
                        method: 'POST',
                        body: chunk,
                        headers: {
                            'Content-Disposition': `attachment; filename="${file.fileName}"`,
                            'Content-Range': `bytes ${startBite}-${endBite}/${file.contentLength}`,
                        },
                        signal: abortController?.signal,
                    });
                    if (uploadRes.status >= 400) {
                        const error = await uploadRes.json();
                        throw new api_1.MaxError(uploadRes.status, error);
                    }
                    if (!uploadData) {
                        uploadData = await uploadRes.json();
                    }
                    file.stream.resume();
                    prevChunk += chunk.length;
                });
                file.stream.on('end', async () => {
                    if (!uploadData) {
                        reject(new Error('Failed to upload'));
                        return;
                    }
                    resolve(uploadData);
                });
                file.stream.on('error', (err) => {
                    reject(err);
                });
            });
        };
        this.uploadFromBuffer = async ({ file, uploadUrl, abortController }) => {
            const formData = new FormData();
            formData.append('data', new Blob([file.buffer]), file.fileName);
            const res = await fetch(uploadUrl, {
                method: 'POST',
                body: formData,
                signal: abortController?.signal,
            });
            return await res.json();
        };
        this.image = async ({ timeout, ...source }) => {
            if ('url' in source) {
                return { url: source.url };
            }
            const fileBlob = await this.getStreamFromSource(source.source);
            return this.upload('image', fileBlob, { timeout });
        };
        this.video = async ({ source, ...options }) => {
            const fileBlob = await this.getStreamFromSource(source);
            return this.upload('video', fileBlob, options);
        };
        this.file = async ({ source, ...options }) => {
            const fileBlob = await this.getStreamFromSource(source);
            return this.upload('file', fileBlob, options);
        };
        this.audio = async ({ source, ...options }) => {
            const fileBlob = await this.getStreamFromSource(source);
            return this.upload('audio', fileBlob, options);
        };
    }
}
exports.Upload = Upload;
