// @maxhub/max-bot-api/dist/core/network/api/modules/messages/api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessagesApi = void 0;
const utils_1 = require("../../../../../utils");
const error_1 = require("../../error");
const base_api_1 = require("../../base-api");
class MessagesApi extends base_api_1.BaseApi {
    constructor() {
        super(...arguments);
        this.get = async ({ ...query }) => {
            return this._get('messages', {
                query,
            });
        };
        this.getById = async ({ message_id }) => {
            return this._get('messages/{message_id}', {
                path: { message_id },
            });
        };
        this.send = async ({ chat_id, user_id, disable_link_preview, ...body }) => {
            try {
                return await this._post('messages', {
                    body,
                    query: { chat_id, user_id, disable_link_preview },
                });
            }
            catch (err) {
                if (err instanceof error_1.MaxError) {
                    if (err.code === 'attachment.not.ready') {
                        console.log('Attachment not ready');
                        await (0, utils_1.sleep)(1000);
                        return this.send({
                            chat_id, user_id, disable_link_preview, ...body,
                        });
                    }
                }
                throw err;
            }
        };
        this.edit = async ({ message_id, ...body }) => {
            return this._put('messages', {
                query: { message_id },
                body,
            });
        };
        this.delete = async ({ ...query }) => {
            return this._delete('messages', {
                query,
            });
        };
        this.answerOnCallback = async ({ callback_id, ...body }) => {
            return this._post('answers', {
                query: { callback_id },
                body,
            });
        };
    }
}
exports.MessagesApi = MessagesApi;
