// @maxhub/max-bot-api/dist/core/network/api/client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createClient = void 0;
const debug_1 = __importDefault(require("debug"));
const debug = (0, debug_1.default)('one-me:client');
const defaultOptions = {
    baseUrl: 'https://botapi.max.ru',
};
const createClient = (token, options = {}) => {
    const { baseUrl } = { ...defaultOptions, ...options };
    const call = async ({ method, options: callOptions }) => {
        const httpMethod = callOptions.method || 'GET';
        debug(`Call method ${httpMethod} /${method}`, JSON.stringify(callOptions, null, 2));
        if (!token) {
            return {
                status: 401,
                data: {
                    code: 'verify.token',
                    message: 'Empty access_token',
                },
            };
        }
        const url = new URL(buildUrl(method, token, callOptions.path), baseUrl);
        Object.keys(callOptions.query ?? {}).forEach((param) => {
            const value = callOptions.query?.[param];
            if (!value)
                return;
            url.searchParams.set(param, value.toString());
        });
        const init = { ...getResponseInit(callOptions?.body), method: httpMethod };
        const res = await fetch(url.href, init);
        if (res.status === 401) {
            return {
                status: 401,
                data: {
                    code: 'verify.token',
                    message: 'Invalid access_token',
                },
            };
        }
        return {
            status: res.status,
            data: await res.json(),
        };
    };
    return { call };
};
exports.createClient = createClient;
const getResponseInit = (body) => {
    if (!body)
        return {};
    return {
        body: JSON.stringify(body),
        headers: {
            'content-type': 'application/json',
        },
    };
};
const buildUrl = (baseUrl, token, path) => {
    let url = baseUrl;
    if (path) {
        Object.keys(path)?.forEach((key) => {
            const regexp = new RegExp(`{${key}}`, 'g');
            const value = path[key].toString();
            url = url.replace(regexp, value);
        });
    }
    return `${url}?access_token=${token}`;
};
